examples/ConfigurationStoreBackup/BackupAppConfigurationStore.cs

```diff
@@ -0,0 +1,163 @@
+                    decodedMessage = Encoding.UTF8.GetString(Convert.FromBase64String(message.MessageText));
+                    JToken dataObject = JObject.Parse(decodedMessage).Property("data")?.Value;
+                    if (dataObject != null && dataObject["key"] != null)
+                    {
+                        string key = dataObject["key"].ToString();
+                        string label = dataObject["label"]?.ToString();
+                        updatedKeyLabels.Add(new KeyLabel(key, label));
+                    }
+                }
+                catch (FormatException)
+                {
+                    log.LogInformation($"Queue message in invalid Base64 format will be ignored.\nMessage: {message.MessageText}");
+                }
+                catch (JsonReaderException)
+                {
+                    log.LogInformation($"Queue message in invalid JSON format will be ignored.\nMessage: {decodedMessage}");
+                }
+            }
+            return updatedKeyLabels;
+        }
+
+        private static async Task<bool> BackupKeyValuesAsync(HashSet<KeyLabel> updatedKeyLabels, ILogger log)
+        {
+            bool isBackupSuccessful = false;
+            try
+            {
+                // Read all settings from primary store and update secondary store.
+                await foreach (ConfigurationSetting setting in primaryAppConfigClient.GetConfigurationSettingsAsync(new SettingSelector()).ConfigureAwait(false))
+                {
+                    KeyLabel primaryStoreKeyLabel = new KeyLabel(setting.Key, setting.Label);
+                    if (updatedKeyLabels.Contains(primaryStoreKeyLabel))
+                    {
+                        // Current setting retrieved from primary store needs to be updated in secondary store.
+                        await secondaryAppConfigClient.SetConfigurationSettingAsync(setting).ConfigureAwait(false);
+                        log.LogInformation($"Successfully updated key: {setting.Key} label: {setting.Label}");
+
+                        updatedKeyLabels.Remove(primaryStoreKeyLabel);
+                        if (updatedKeyLabels.Count == 0)
+                        {
+                            break;
+                        }
+                    }
+                }
+
+                // Delete key labels still present in updatedKeyLabels from secondary store.
+                foreach (var keyLabel in updatedKeyLabels)
+                {
+                    await secondaryAppConfigClient.DeleteConfigurationSettingAsync(keyLabel.Key, keyLabel.Label).ConfigureAwait(false);
+                }
+                isBackupSuccessful = true;
+            }
+            catch (RequestFailedException exception)
+            {
+                log.LogError($"Service request failed from the server.\nStatus: {exception.Status}\nError Message: {exception.Message}");
```

Comment:
I hope we can have clearer error message here. How about "Request to Azure App Configuration failed...."

Maybe we should include call stack. It's hard to tell whether it's request to the primary or the secondary failed. Is there an easy way?

Reply:
I looked at all the properties of this exception. There is no way to tell which config store is throwing the exception. We can add call stack as it can help isolate the call that is throwing the exception.
This is all the data we get in the exception for reference:

Status: 409
Error Message: The setting is read only
 Status: 409 (Conflict)

 Content:
 {"type":"[https://azconfig.io/errors/key-locked](https://azconfig.io/errors/key-locked)","title":"Modifing key 'abc' is not allowed","name":"abc","detail":"The key is read-only. To allow modification unlock it first.","status":409}

 Headers:
 Server: nginx/1.16.1
 Date: Wed, 22 Apr 2020 18:50:11 GMT
 Transfer-Encoding: chunked
 Connection: keep-alive
 x-ms-request-id: 9354d085-cf8c-4486-b8c5-b09b66ba6add
 x-ms-client-request-id: d5baa60f-9008-4c44-ad99-344ad459c352
 x-ms-correlation-request-id: 9354d085-cf8c-4486-b8c5-b09b66ba6add
 Access-Control-Allow-Origin: *
 Access-Control-Allow-Credentials: true
 Access-Control-Expose-Headers: DNT, X-CustomHeader, Keep-Alive, User-Agent, X-Requested-With, If-Modified-Since, Cache-Control, Content-Type, Authorization, x-ms-client-request-id, x-ms-useragent, x-ms-content-sha256, x-ms-date, host, Accept, Accept-Datetime, Date, If-Match, If-None-Match, Sync-Token, x-ms-return-client-request-id, ETag, Last-Modified, Link, Memento-Datetime, x-ms-retry-after, x-ms-request-id, WWW-Authenticate
 Strict-Transport-Security: max-age=15724800; includeSubDomains
 Content-Type: application/problem+json; charset=utf-8

Stack Trace:    at Azure.Data.AppConfiguration.ConfigurationClient.DeleteConfigurationSettingAsync(String key, String label, MatchConditions requestOptions, CancellationToken cancellationToken)
   at Azure.Data.AppConfiguration.ConfigurationClient.DeleteConfigurationSettingAsync(String key, String label, CancellationToken cancellationToken)
   at ConfigurationStoreBackup.BackupAppConfigurationStore.BackupKeyValuesAsync(HashSet`1 updatedKeyLabels, ILogger log) in D:\SampleProjects\bcdr\BackupFromStorage\BackupFromStorage\BackupAppConfigurationStore.cs:line 149

Data: System.Collections.ListDictionaryInternal
ErrorCode : 
HelpLink : 
HResult : -2146233088
InnerException :
Source: Azure.Data.AppConfiguration
TargetSite: Void MoveNext()



---
In reply to: [412703340](https://github.com/Azure/AppConfiguration/pull/311#discussion_r412703340) [](ancestors = 412703340)

Reply:
I would just dump out this whole thing. The request-id is in it. At least, customers can contact us if they think this is unexpected.

Reply:
Request ID is already included in exception.Message. 

---
In reply to: [413289512](https://github.com/Azure/AppConfiguration/pull/311#discussion_r413289512) [](ancestors = 413289512)